onnxruntime/core/providers/xnnpack/math/gemm.cc

```diff
@@ -0,0 +1,193 @@
+// Copyright(c) Microsoft Corporation.All rights reserved.
+// Licensed under the MIT License.
+
+#include "gemm.h"
+#include "core/framework/transpose_helper.h"
+#include "core/providers/utils.h"
+
+namespace onnxruntime {
+namespace xnnpack {
+
+bool Gemm::IsGemmOnnxNodeSupported(const NodeUnit& node_unit, const GraphViewer& graph) {
+  bool supported = false;
+  const onnxruntime::Node& node = node_unit.GetNode();
+
+  // use do {} while(false) so it's easier to set a breakpoint on the return
+  do {
+    ConstPointerContainer<std::vector<NodeArg*>> input_defs = node.InputDefs();
+
+    const auto alpha = node.GetAttributes().find("alpha");
+    if (!(*alpha).second.has_f() && (*alpha).second.f() != 1.0) break;
+
+    const auto beta = node.GetAttributes().find("beta");
+    if (!(*beta).second.has_f() && (*beta).second.f() != 1.0) break;
+
+    const NodeArg* A_arg = input_defs[0];
+    const NodeArg* B_arg = input_defs[1];
+    const NodeArg* C_arg = input_defs.size() <= 2 ? nullptr : input_defs[2];
+
+    // Right now assuming C matrix is exists
+    if (!C_arg) break;
+
+    // we only support float currently
+    const auto* A_type = A_arg->TypeAsProto();
+
+    if (A_type == nullptr || 
+        A_type->tensor_type().elem_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT) {
+      break;
+    }
+
+    // B & C matrices must be constant
+    if (!graph.IsConstantInitializer(B_arg->Name(), true)) {
+      break;
+    }
+
+    if (input_defs[2]->Exists() && !graph.IsConstantInitializer(C_arg->Name(), true)) {
+      break;
+    }
+
+    // making sure we are dealing with MatMul
+    const ONNX_NAMESPACE::TensorShapeProto* A_shape = A_arg->Shape();
+    const ONNX_NAMESPACE::TensorShapeProto* B_shape = B_arg->Shape();
+    const ONNX_NAMESPACE::TensorShapeProto* C_shape = C_arg->Shape();
+
+    if (!A_shape || A_shape->dim_size() >= 3) {
+      break;
+    }
+
+    if (!B_shape || B_shape->dim_size() >= 3) {
+      break;
+    }
+
+    if (!C_shape || C_shape->dim_size() >= 3) {
+      break;
+    }
+    
+    if (C_shape->dim(0).dim_value() != B_shape->dim(1).dim_value() && C_shape->dim(0).dim_value() != B_shape->dim(0).dim_value()){
```

Comment:
Wondering if this is too strict as C only has to be broadcast-able. We use GemmBroadcastBias in the CPU implementation so some logic from there can probably be borrowed. Can add a comment in the code and do that as a separate PR if necessary.